### PR TITLE
Ammo HUD

### PR DESCRIPTION
**Pull request by DominikPanic**
:cl: LaKiller8
add: Implements an Ammo HUD to the vast majority of weapons, which becomes visible upon wielding, shows your current amount of ammo, and the ammo type of the current bullet. Once it reaches 0 rounds it will warns you by flashing. It can be toggled for every gun.
imageadd: Ammo HUD sprites made by ArachnidNexus.
refactor: Cleans up the energy weapons code, which now use batteries for an energy source.
balance: As a side effect, pred rifles no longer self-chages for now.
/:cl:

Implements the Ammo HUD sprites made by ArachnidNexus

dreamseeker_2018-12-02_04-37-22

### CODE REVIEW

File: code/_onclick/hud/screen_objects.dm [hunk 1]
```diff
@@ -617,3 +617,121 @@
...
+
+	if(G.current_mag)
+		rounds += G.current_mag.current_rounds
+
+	if(G.in_chamber && !istype(G, /obj/item/weapon/gun/launcher) && !istype(G, /obj/item/weapon/gun/revolver) && !istype(G, /obj/item/weapon/gun/smartgun) && !istype(G, /obj/item/weapon/gun/energy))
+		rounds++
+
+	if(istype(G, /obj/item/weapon/gun/energy))
+		var/obj/item/weapon/gun/energy/E = G
+		rounds += (E.cell.charge / E.charge_cost)
```
Comment (Rohesie) [inline code/_onclick/hud/screen_objects.dm]: Do an
```
else if(G.in_chamber && !istype(G, /obj/item/weapon/gun/launcher) && !istype(G, /obj/item/weapon/gun/revolver) && !istype(G, /obj/item/weapon/gun/smartgun))
	rounds++
```
here, avoiding the double-check.
Or even better, add a variable in those guns to account for this, instead of checking every type.

Comment (Rohesie) [reply code/_onclick/hud/screen_objects.dm]: Also a safe check for a division-by-zero would be good.

File: code/_onclick/hud/screen_objects.dm [hunk 2]
```diff
@@ -617,3 +617,121 @@
...
+		overlays.Cut()
+		remove_hud()
+		to_chat(user, "Something fucky with your ammo fam.")
+		return
+
+	overlays.Cut()
+
+	rounds = num2text(rounds)
+
+	to_chat(user, "[rounds] rounds")
```
Comment (Rohesie) [inline code/_onclick/hud/screen_objects.dm]: Debug?

File: code/_onclick/hud/screen_objects.dm [hunk 3]
```diff
@@ -617,3 +617,89 @@
...
+
+	var/obj/item/weapon/gun/G = user.get_active_hand()
+
+	if(!G || !G.has_ammo_counter())
+		return
+
+	if(user.client.screen.Find(src))
+		return
+	else
+		user.client.screen += src
```
Comment (ninjanomnom) [inline code/_onclick/hud/screen_objects.dm]: Just add it, byond makes sure the list is unique

File: code/_onclick/hud/screen_objects.dm [hunk 4]
```diff
@@ -617,3 +617,89 @@
...
+
+/obj/screen/ammo/proc/remove_hud(var/mob/user)
+	if(!user?.client)
+		return
+
+	if(!user.client.screen.Find(src))
+		return
+	else
+		user.client.screen -= src
```
Comment (ninjanomnom) [inline code/_onclick/hud/screen_objects.dm]: Same as above, this whole proc could be `user?.client?.screen -= src`

File: code/_onclick/hud/screen_objects.dm [hunk 5]
```diff
@@ -617,3 +617,89 @@
...
+
+/obj/screen/ammo/proc/update_hud(var/mob/user)
+	if(!user?.client || !user.client.screen.Find(src))
```
Comment (ninjanomnom) [inline code/_onclick/hud/screen_objects.dm]:
```suggestion
	if(!user?.client?.screen.Find(src))
```

File: code/_onclick/hud/screen_objects.dm [hunk 6]
```diff
@@ -617,3 +617,89 @@
...
+			overlays += image('icons/mob/ammoHUD.dmi', src, "t[rounds[1]]")
+		if(3)
+			overlays += image('icons/mob/ammoHUD.dmi', src, "o[rounds[3]]")
+			overlays += image('icons/mob/ammoHUD.dmi', src, "t[rounds[2]]")
+			overlays += image('icons/mob/ammoHUD.dmi', src, "h[rounds[1]]")
+		else //"0" is still length 1 so this means it's over 999
+			overlays += image('icons/mob/ammoHUD.dmi', src, "o9")
+			overlays += image('icons/mob/ammoHUD.dmi', src, "t9")
+			overlays += image('icons/mob/ammoHUD.dmi', src, "h9")
+			return
```
Comment (ninjanomnom) [inline code/_onclick/hud/screen_objects.dm]: This return isnt needed

File: code/game/objects/items/explosives/grenades/marines.dm [hunk 7]
```diff
@@ -48,77 +49,85 @@
...
+	spawn(0)
+		explosion(loc, -1, -1, 4)
+		cdel(src)
+	return
 
-	prime()
-		spawn(0)
-			explosion(loc, -1, -1, 4)
-			cdel(src)
-		return
```
Comment (spookydonut) [inline code/game/objects/items/explosives/grenades/marines.dm]: don't do this unless you're doing a reformatting pr because it makes it more time consuming to review

File: code/modules/cm_marines/equipment/weapons.dm [hunk 8]
```diff
@@ -74,19 +74,30 @@
...
 		else
-			to_chat(user, "Your reloading was interrupted!")
-			playsound(src,'sound/machines/buzz-two.ogg', 25, 1)
-			reloading = FALSE
-			return
-	else
-		if(autoload_check(user, reload_duration, mygun, src))
-			reload(user, mygun, TRUE)
+			user.visible_message("[user.name]'s powerpack servos begin automatically feeding an ammo belt into the M56 Smartgun.","The powerpack servos begin automatically feeding a fresh ammo belt into the M56 Smartgun.")
+		if(!automatic)
```
Comment (spookydonut) [inline code/modules/cm_marines/equipment/weapons.dm]: There's a bug here somewhere which stops smartguns reloading.

File: code/modules/cm_marines/super_secret.dm [hunk 9]
```diff
@@ -56,12 +56,12 @@
 							for(var/mob/living/carbon/human/H in range(12))
 								if(!H || isnull(H) || H.stat) continue
 								to_chat(H, "<font size=8 color=red><b>NO<br><br>BUGS</b></font>")
-							new /obj/item/weapon/gun/launcher/rocket/nobugs(get_turf(src.loc))
```
Comment (spookydonut) [inline code/modules/cm_marines/super_secret.dm]: unrelated to the PR, should be removed.

File: code/modules/cm_preds/yaut_items.dm [hunk 10]
```diff
@@ -493,7 +493,6 @@
 
 		var/obj/item/weapon/gun/energy/plasma_caster/W = new(usr)
 		usr.put_in_active_hand(W)
-		W.source = src
```
Comment (spookydonut) [inline code/modules/cm_preds/yaut_items.dm]: why

File: code/modules/projectiles/updated_projectiles/ammo_datums.dm [hunk 11]
```diff
@@ -1480,20 +1552,20 @@
 	G.damtype = "fire"
 	G.SetLuminosity(G.brightness_on)
 
-/datum/ammo/rocket/nobugs
-	name = "\improper NO BUGS rocket"
+/datum/ammo/rocket/toy
```
Comment (spookydonut) [inline code/modules/projectiles/updated_projectiles/ammo_datums.dm]: unrelated change, should be separate pr

File: code/modules/projectiles/updated_projectiles/gun_system.dm [hunk 12]
```diff
@@ -95,23 +95,23 @@
...
-		..()					//This only affects guns you can get from vendors for now. Special guns spawn with their own things regardless.
-		base_gun_icon = icon_state
-		attachable_overlays = list("muzzle", "rail", "under", "stock", "mag", "special")
-		if(current_mag)
-			if(spawn_empty && !(flags_gun_features & GUN_INTERNAL_MAG)) //Internal mags will still spawn, but they won't be filled.
-				current_mag = null
-				update_icon()
-			else
-				current_mag = new current_mag(src, spawn_empty ? 1 : 0)
-				ammo = current_mag.default_ammo ? ammo_list[current_mag.default_ammo] : ammo_list[/datum/ammo/bullet] //Latter should never happen, adding as a precaution.
```
Comment (spookydonut) [inline code/modules/projectiles/updated_projectiles/gun_system.dm]: same here

File: code/modules/projectiles/updated_projectiles/guns/energy.dm [hunk 13]
```diff
@@ -87,34 +113,12 @@
...
-	verbs -= /obj/item/weapon/gun/verb/empty_mag
-
-
-
-/obj/item/weapon/gun/energy/plasmarifle/Dispose()
 	. = ..()
-	processing_objects.Remove(src)
-
-
-/obj/item/weapon/gun/energy/plasmarifle/process()
```
Comment (spookydonut) [inline code/modules/projectiles/updated_projectiles/guns/energy.dm]: why did you remove the self-charging?  undocumented change

File: code/modules/projectiles/updated_projectiles/guns/flamer.dm [hunk 14]
```diff
@@ -142,30 +144,27 @@
 	var/burnlevel
 	var/burntime
 	var/fire_color = "red"
-	switch(current_mag.caliber)
-		if("UT-Napthal Fuel") //This isn't actually Napalm actually
```
Comment (spookydonut) [inline code/modules/projectiles/updated_projectiles/guns/flamer.dm]: unrelated changes

File: code/_onclick/hud/screen_objects.dm [hunk 15]
```diff
@@ -617,3 +617,79 @@
...
+			flick("[hud_state_empty]_flash", F)
+			spawn(20)
+				user.client.screen -= F
+				cdel(F)
+				overlays += empty
+	else
+		warned = FALSE
+		overlays += image('icons/mob/ammoHUD.dmi', src, "[hud_state]")
+
+	rounds = num2text(rounds)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/_onclick/hud/screen_objects.dm]: The reason I want it as a string is so I can later easily add the different digits:
```"o[rounds[3]]"
"t[rounds[2]]"
"h[rounds[1]]"```

File: code/game/objects/items/explosives/grenades/marines.dm (hunk 7 above)
Comment (Rohesie) [reply code/game/objects/items/explosives/grenades/marines.dm]: In this case the code hasn't changed, and the code cleanup is positive.

Comment (Rohesie) [reply code/game/objects/items/explosives/grenades/marines.dm]: I'm assuming there are code changes in this file, though.

Comment (DominikPanic) [reply code/game/objects/items/explosives/grenades/marines.dm]: Yes there are.

**Comment by Rohesie** [comment]
Conflicts.

File: code/modules/cm_marines/super_secret.dm (hunk 9 above)
Comment (Rohesie) [reply code/modules/cm_marines/super_secret.dm]: Ideally this should be on its own PR, for sure.

File: code/modules/cm_preds/yaut_items.dm (hunk 10 above)
Comment (DominikPanic) [reply code/modules/cm_preds/yaut_items.dm]: I reworked the energy weapons code to use their own energy sources instead of the badly coded pred bracer one.

File: code/modules/projectiles/updated_projectiles/guns/energy.dm (hunk 13 above)
Comment (DominikPanic) [reply code/modules/projectiles/updated_projectiles/guns/energy.dm]: It was a snowflake mechanic made for preds, not OOP or modular. All energy weapons now use cells properly.

File: code/modules/projectiles/updated_projectiles/guns/flamer.dm (hunk 14 above)
Comment (DominikPanic) [reply code/modules/projectiles/updated_projectiles/guns/flamer.dm]: The switching of the flamer ammo types was bad so I slightly reworked it into something better.

File: code/modules/projectiles/updated_projectiles/gun_system.dm (hunk 12 above)
Comment (DominikPanic) [reply code/modules/projectiles/updated_projectiles/gun_system.dm]: Code cleanup, sorry dad.

File: code/modules/projectiles/updated_projectiles/ammo_datums.dm (hunk 11 above)
Comment (DominikPanic) [reply code/modules/projectiles/updated_projectiles/ammo_datums.dm]: I agree, separating it with my shitty commits wouldn't be easy though. I'll watch for this sort of thing in the future.

File: code/modules/cm_marines/super_secret.dm (hunk 9 above)
Comment (DominikPanic) [reply code/modules/cm_marines/super_secret.dm]: I'll do better in the future.
